packages/hash/api/src/db/postgres/aggregation/sql/aggregations.util.ts

```diff
@@ -0,0 +1,234 @@
+
+export const aggregationsColumnNames = [
+  "aggregation_id",
+  "path",
+  "source_account_id",
+  "source_entity_id",
+  "applied_to_source_at",
+  "applied_to_source_by_account_id",
+  "removed_from_source_at",
+  "removed_from_source_by_account_id",
+];
+
+export const aggregationsColumnNamesSQL = mapColumnNamesToSQL(
+  aggregationsColumnNames,
+);
+
+export type DbAggregationRow = {
+  aggregation_id: string;
+  source_account_id: string;
+  source_entity_id: string;
+  path: string;
+  applied_to_source_at: string;
+  applied_to_source_by_account_id: string;
+  removed_from_source_at: string | null;
+  removed_from_source_by_account_id: string | null;
+};
+
+type DbAggregationWithVersionRow = DbAggregationRow & DbAggregationVersionRow;
+
+/** selects all aggregation versions in the datastore */
+const selectAllAggregationVersions = sql<DbAggregationWithVersionRow>`
+  select
+    ${mapColumnNamesToSQL(aggregationsColumnNames, "aggregations")},
+    ${mapColumnNamesToSQL(
+      aggregationVersionsColumnNames.filter(
+        (column) => !aggregationsColumnNames.includes(column),
+      ),
+      "aggregation_versions",
+    )}
+  from
+    aggregations
+    join aggregation_versions
+    on aggregations.aggregation_id = aggregation_versions.aggregation_id
+`;
+
+/** selects all aggregations in the datastore with their latest version */
+const selectAllLatestVersionsOfAggregations = sql<DbAggregationWithVersionRow>`
+  select distinct on (aggregation_id) *
+  from (${selectAllAggregationVersions}) as all_aggregation_versions
+  order by aggregation_id, updated_at desc
+`;
+
+/** select all aggregations with the latest version before a given timestamp */
+const selectAllLatestVersionsOfAggregationsBeforeTimestamp = (params: {
+  beforeTimestamp: Date;
+}) =>
+  sql<DbAggregationWithVersionRow>`
+    select distinct on (aggregation_id) *
+    from (
+      ${selectAllAggregationVersions}
+      where aggregation_versions.updated_at <= ${params.beforeTimestamp.toISOString()}
+    ) as all_aggregation_versions
+    order by aggregation_id, updated_at desc
+`;
+
+/** selects all versions of a specifc aggregation */
+const selectAllVersionsOfAggregation = (params: {
+  sourceAccountId: string;
+  aggregationId: string;
+}) => sql<DbAggregationWithVersionRow>`
+  with all_aggregations as (${selectAllAggregationVersions})
+  select *
+  from all_aggregations
+  where
+    source_account_id = ${params.sourceAccountId} and aggregation_id = ${params.aggregationId}
+`;
+
+/** selects the latest version of a specific aggregation */
+export const selectLatestVersionOfAggregation = (params: {
+  sourceAccountId: string;
+  aggregationId: string;
+}) => sql<DbAggregationWithVersionRow>`
+  with all_aggregation_versions as (${selectAllVersionsOfAggregation(params)})
+  select distinct on (aggregation_id) *
+  from all_aggregation_versions
+  order by aggregation_id, updated_at desc
+`;
+
+export const selectAllAggregationsWithSourceEntity = (params: {
```

Comment:
Missing a docstring (maybe not important but stands out as the only one)

Reply:
Addressed in https://github.com/hashintel/hash/pull/484/commits/6af383a178fa99fe9bb62cfb5571a1abf683da00